### ISSUE TITLE
Lesson 3: Animation

### ISSUE THREAD

**Issue by runarberg** [K0rston/Pong#3]
Now we have drawn simple shapes onto the canvas it is time to learn about animation. The goal is to make the ball and paddles move somehow. Here is an example:

![animation](https://github.com/K0rston/Pong/assets/3984600/6c4014a9-9a91-494a-be0c-419a9035cf56)

An animation is basically just a series of drawings shown one at a time in quick succession. We will ask the browser when it is going to redraw to the screen with a function called [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) (so basically a way to listen to the display’s framerate). Then we will paint over our previous drawing and redraw with the entities moved slightly.

From there we will go on to create something called a game loop, which is basically a loop that runs ones per frame. A game loop is pretty much at the heart of every game engine.

An example of a simple game loop can look something like this:

```js
function nextAnimationFrame() {
  return new Promise((resolve) => {
    window.requestAnimationFrame(resolve);
  });
}

// Create an infinite loop.
while (true) {
  await nextAnimationFrame();
  // We are here only once per frame. Run the game logic.
  collectInputs();
  updateState();
  render();
}
```

For now we will only focus on the render part of the game loop.

---

Aside: If you are interested I recorded the gif above with a web api called Media Capture API: https://developer.mozilla.org/en-US/docs/Web/API/Media_Capture_and_Streams_API

<details>
<summary>Expand to see code example:</summary>

```js
function record(canvas) {
  const videochunks = [];
  const stream = canvas.captureStream();
  const rec = new MediaRecorder(stream);

  rec.ondataavailable = (event) => {
    videochunks.push(event.data);
  };

  rec.onstop = () => {
    const blob = new Blob(videochunks, { type: "video/webm" });
    const file = new File([blob], "animation.webm");
    const href = URL.createObjectURL(file);
    const anchor = document.createElement("a");

    anchor.href = href;
    anchor.download = file.name;
    anchor.textContent = file.name;

    document.body.appendChild(anchor);
  };

  return rec;
}

```
</details>

**Comment by runarberg** [comment]
Below is a diff of the final step to get the ball moving. The red lines (prepended with `-`) are lines you should remove, and the green lines (prepended with `+`) you should add. The entire diff is also available in  92f2146e.

Starting with the `main` function we create a `ball` object:

```js
const ball = {
  x: 250,
  y: 250,
};
```

which is basically just a container around an x and y coordinates. We will alter the x and y coordinates (or rather just the x coordinate for now) on this object with each subsequent frame, by mutating the object inside the game loop:

```js
  ball.x += 1;
```

But to we need to draw this updated state, and for that we need to pass the ball into the draw function:

```diff
- draw(ctx);
+ draw(ctx, ball);
```

So now the `main` function should look something like this:

```js
async function main() {
  const canvas = document.getElementById("game");

  if (!(canvas instanceof HTMLCanvasElement)) {
    return;
  }

  const ctx = canvas.getContext("2d");

  // Create an object to hold the x and y coordinates of the ball.
  const ball = {
    x: 250,
    y: 250,
  };

  for await (const frame of animationFrames()) {
    // Draw the ball with the canvas context.
    draw(ctx, ball);

    // Move the ball down one pixel
    ball.y += 1;

    // Exit the loop after 200 frames
    if (frame >= 200) {
      break;
    }
  }
}
```

That final bit in the main loop `if (frame >= 200) { break; }` is non-essential. It will stop the loop (i.e. freeze the animation) after 200 frames. If you omit it—which you should try—the ball will just keep going down until it exits the canvas (and even then it will continue to go down).

But for this to work the `draw` function needs to accept the ball as a parameter:

```diff
  /**
   * @param {CanvasRenderingContext2D} ctx
+  * @param {{x: number, y: number}} ball
   * @returns {void}
   */
- function draw(ctx) {
+ function draw(ctx, ball) {
```

And finally instead of drawing a static x=250 and y=250 we draw whatever we pass in as the ball:

```diff
  // Ball
  ctx.beginPath();
  ctx.fillStyle = "black";
- ctx.arc(250, 250, 7.5, 0, 2 * Math.PI);
+ ctx.arc(ball.x, ball.y, 7.5, 0, 2 * Math.PI);
  ctx.fill();
```

So the final draw function will look something like this:

```js
/**
 * @param {CanvasRenderingContext2D} ctx
 * @param {{x: number, y: number}} ball
 * @returns {void}
 */
function draw(ctx, ball) {
  // Clear the canvas with a white background
  ctx.fillStyle = "white";
  ctx.rect(0, 0, 500, 500);
  ctx.fill();

  // Ball
  ctx.beginPath();
  ctx.fillStyle = "black";
  ctx.arc(ball.x, ball.y, 7.5, 0, 2 * Math.PI);
  ctx.fill();

  // Paddle 1
  ctx.beginPath();
  ctx.fillStyle = "black";
  ctx.rect(100, 20, 80, 15);
  ctx.fill();

  // Paddle 2
  ctx.beginPath();
  ctx.fillStyle = "black";
  ctx.rect(320, 465, 80, 15);
  ctx.fill();
}
```

**Comment by runarberg** [comment]
@K0rston Try to write this either in your jsfiddle or on your machine at home (you can copy-paste the files ([available here](https://github.com/K0rston/Pong/tree/92f2146e2e71dc26df4d45c343b122ab1c3e6159)) into a directory and open `index.html` with your browser). If you run into trouble you can reply below and tag my handle.

If you do this and see the ball move, you can do similar stuff for your two paddles as you did with the ball. Additionally:
* Try to make the ball move faster (hint you can increase the number of pixels it moves on every frame from 1 to say 5).
* Try to make the ball move diagonally by altering the y coordinate as well as the x.
* Try to make the ball change direction after 200 frames.
* Try to make the ball and the two paddles all move at the same time.

If this all goes well you can try to mimic the animation I show at the top of this issue.
